src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -491,6 +491,9 @@ public IThreadPoolWorkItem Dequeue(ThreadPoolWorkQueueThreadLocals tl, ref bool
 
         internal static bool Dispatch()
         {
+            // Put threadpool thread on default context
+            ExecutionContextSwitcher.SetDefaults();
```

Comment:
> Why not leave it as null (and allow restore to null) and also check for null (maybe first) when checking for default? Is it significantly better to initialize to Default and check for only that?

`SetDefaults` does check if its already default before assigning it. `null` is a bit of a weird one; if you `Capture` it `?? Deafult`; however if you are inspecting it directly then it counts a flow suppressed.

But mainly it doubles the number of branches and is whether the `JIT_XxxWriteBarrier` is called once per item executed or once per threadpool thread lifetime. 

As running on the default context is currently baked into `QueueUserWorkItemCallbackDefaultContext` and ATMB.`MoveNextRunner.RunWithDefaultContext` and with this change into `Task.ExecuteWithThreadLocal` it makes sense to just leave the threadpool thread on the default context; rather than moving to null and always restoring the Default context; especially since non-default contexts will always restore to previous; whether that is null or Default.

So with this change the 3 definitely run on default context paths call the new method `RunDefaultContext` and don't bother passing context as an argument.

Other flows which may be on the default context call `Run` as they were doing before, still taking the previous fast path, but receiving the advantage that the Threadpool is already on the default context rather than null so the context doesn't need to be assigned to Default or restored back to null.

Its just introducing an even faster path when the Run context is known to be Default upfront.